### PR TITLE
 consent pages (ylp 506 and ylp 507) 

### PR DESCRIPTION
**Pull request by Nas84**
maybe waiting the #505 could be interesting and require also de-serialisation function of the user connected from the storage

### CODE REVIEW

**Review by yvon-dblg — COMMENTED** [review]
Partial review
I will continue later

File: packages/yourloops/components/consent.tsx [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+  const { t, i18n } = useTranslation("yourloops");
+  const auth = useAuth();
+  const historyHook = useHistory();
+  const classes = style();
+  const [terms, setTerms] = React.useState(false);
+  const [privacyPolicy, setPrivacyPolicy] = React.useState(false);
+  const [error, setError] = React.useState(false);
+  const [helperText, setHelperText] = React.useState("");
+
+  const linkTermsText = t("Terms of Use");
```
Comment (yvon-dblg) [inline packages/yourloops/components/consent.tsx:113]:
```suggestion
  const linkTermsText = t("terms-and-conditions");
```

File: packages/yourloops/components/consent.tsx [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+  const linkTerms = DiabeloopUrl.getTermsLink(i18n.language);
+  const privacyPolicyText = t("Privacy Policy");
```
Comment (yvon-dblg) [inline packages/yourloops/components/consent.tsx:115]:
```suggestion
  const privacyPolicyText = t("footer-link-url-privacy-policy");
```

File: packages/yourloops/components/consent.tsx [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    if (valideForm()) {
+      // api call
+      const user = auth?.user;
+      if (user?.profile !== undefined && user?.profile !== null) {
+        const now = new Date().toISOString();
+        user.profile.termsOfUse = { acceptanceTimestamp: now, isAccepted: terms };
+        user.profile.privacyPolicy = { acceptanceTimestamp: now, isAccepted: privacyPolicy };
+        auth.updateProfile(user.profile);
+        historyHook.push(props.destinationPath);
+      }
```
Comment (yvon-dblg) [inline packages/yourloops/components/consent.tsx:156]: We should update the profile even if it not exists yet.
I know, this should not happen.

```suggestion
      const user = auth.user;
      if (user === null) {
        throw new Error("User must be logged-in");
      }
      const now = new Date().toISOString();
      const updatedProfile = _.cloneDeep(user.profile ?? {}) as Profile;
      updatedProfile.termsOfUse = { acceptanceTimestamp: now, isAccepted: terms };
      updatedProfile.privacyPolicy = { acceptanceTimestamp: now, isAccepted: privacyPolicy };
      auth.updateProfile(updatedProfile).catch((reason: unknown) => {
        console.error(reason);
      }).finally(() => {
        historyHook.push(props.destinationPath);
      });
```

Of course some imports are missing with my suggestion.

File: packages/yourloops/components/consent.tsx [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    resetFormState();
+  };
+
+  const valideForm = (): boolean => {
+    if (terms && privacyPolicy) {
+      return true;
+    }
+
+    setError(true);
+    setHelperText("required");
```
Comment (yvon-dblg) [inline packages/yourloops/components/consent.tsx:137]:
```suggestion
    setHelperText(t("required"));
```

File: packages/yourloops/components/consent.tsx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+                paddingBottom: "1em",
+              }}>
+              <img
+                src={brandingLogo}
+                style={{
+                  height: "60px",
+                  marginLeft: "auto",
+                  marginRight: "auto",
+                }}
+                alt={t("Login Branding Logo")}
```
Comment (yvon-dblg) [inline packages/yourloops/components/consent.tsx:183]:
```suggestion
                alt={t("alt-img-logo")}
```

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
Review in progress

File: packages/yourloops/components/consent.tsx (hunk 4 above)
Comment (Nas84) [reply packages/yourloops/components/consent.tsx:140]: not needed is it is done here:
`                  <FormHelperText className={classes.FormHelperText}>
                    {t(helperText)}
                  </FormHelperText>`

File: packages/yourloops/lib/auth/user.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  shouldRenewConsent(): boolean {
+    if (this?.profile?.termsOfUse === undefined || this?.profile?.termsOfUse === null) {
+      return true;
+    }
+
+    if (this?.profile?.privacyPolicy === undefined || this?.profile?.privacyPolicy === null) {
+      return true;
+    }
+
+    return (
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:93]: Wrong check here !
```suggestion
    return !(
```

File: packages/yourloops/lib/auth/user.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+   * Check If the user should renew is consent.
+   */
+  shouldRenewConsent(): boolean {
+    if (this?.profile?.termsOfUse === undefined || this?.profile?.termsOfUse === null) {
+      return true;
+    }
+
+    if (this?.profile?.privacyPolicy === undefined || this?.profile?.privacyPolicy === null) {
+      return true;
+    }
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:91]:
```suggestion
    if (this.shouldAcceptConsent()) {
      return true;
    }
```
Would have been prettier.

File: packages/yourloops/lib/auth/user.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  checkConsent(consent: Consent): boolean {
+
+    if (typeof consent.acceptanceTimestamp === "string") {
+      // A `null` is fine here, because `new Date(null).valueOf() === 0`
+      const acceptDate = new Date(consent.acceptanceTimestamp);
+      if (!Number.isFinite(acceptDate)) {
+        // if acceptDate is not a valid formatted date string, get user to re-accept terms
+        return false;
+      }
+      return acceptDate.valueOf() > 0 && this.latestConsentChangeDate >= acceptDate;
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:112]:
```suggestion
      return acceptDate > 0 && this.latestConsentChangeDate.valueOf() < acceptDate;
```

File: packages/yourloops/lib/auth/user.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  /**
+   * Check the given consent against the lastest consent publication date
+   * @param consent
+   * @returns true if the lastest consent date is greater than the given consent
+   */
+  checkConsent(consent: Consent): boolean {
+
+    if (typeof consent.acceptanceTimestamp === "string") {
+      // A `null` is fine here, because `new Date(null).valueOf() === 0`
+      const acceptDate = new Date(consent.acceptanceTimestamp);
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:107]:
```suggestion
      const acceptDate = new Date(consent.acceptanceTimestamp).valueOf();
```

File: packages/yourloops/lib/auth/user.ts [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      if (!Number.isFinite(acceptDate)) {
+        // if acceptDate is not a valid formatted date string, get user to re-accept terms
+        return false;
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:110]:
```suggestion
        return true;
```

File: packages/yourloops/lib/auth/user.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      }
+      return acceptDate.valueOf() > 0 && this.latestConsentChangeDate >= acceptDate;
+    }
+
+    return false;
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:115]:
```suggestion
    return true;
```

File: packages/yourloops/lib/auth/user.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+   * Check the given consent against the lastest consent publication date
+   * @param consent
+   * @returns true if the lastest consent date is greater than the given consent
+   */
+  checkConsent(consent: Consent): boolean {
+
+    if (typeof consent.acceptanceTimestamp === "string") {
+      // A `null` is fine here, because `new Date(null).valueOf() === 0`
+      const acceptDate = new Date(consent.acceptanceTimestamp);
+      if (!Number.isFinite(acceptDate)) {
```
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:108]: Checking if the number is finite against a Date !
Should always be false

File: packages/yourloops/lib/auth/user.ts (hunk 8 above)
Comment (yvon-dblg) [inline packages/yourloops/lib/auth/user.ts:112]: I'm not fan comparing date directly.
I would prefer compare numbers.

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
The consent check is not good.
This is a required check !
For the rest, they are more or less mandatory

File: packages/yourloops/components/routes.tsx [hunk 13]
```diff
@@ -33,19 +33,15 @@ import CssBaseline from "@material-ui/core/CssBaseline";
...
 
 export const PublicRoute = (props: RouteProps): JSX.Element => {
-  const { isLoggedIn, user } = useAuth();
 
-  return isLoggedIn() ? (
-    <Redirect to={{ pathname: getURLPrefixFromUser(user), state: { from: props.location } }} />
-  ) : (
-    <ThemeProvider theme={publicRoutesTheme}>
+  return (
+    <ThemeProvider theme={externalTheme}>
```
[the remark this answers is not in the document]
Comment (Nas84) [reply packages/yourloops/components/routes.tsx:44]: This is not out of scope. We do not have the choice, if we don't change that the consent page will not appears. this behaviour breaks the history stack.
 I think it is a wrong  design choice to we have several place that contain the logic that push the user home page.

File: packages/yourloops/lib/auth/api.ts [hunk 14]
```diff
@@ -43,6 +44,14 @@ import { Session, UpdateUser } from "./models";
 const log = bows("Auth API");
 const failedLoginCounter = new Map<string, number>();
 
+function format(user: any): User {
```
[the remark this answers is not in the document]
Comment (Nas84) [reply packages/yourloops/lib/auth/api.ts:47]: will be done later

File: packages/yourloops/lib/auth/user.ts (hunk 12 above)
Comment (Nas84) [reply packages/yourloops/lib/auth/user.ts:108]: i took your proposition ! 😆

File: packages/yourloops/lib/auth/user.ts [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+  role!: UserRoles;
+  emails?: string[];
+  emailVerified?: boolean;
+  profile?: Profile | null;
+  settings?: Settings | null;
+  preferences?: Preferences | null;
+  medicalData?: MedicalData | null;
+  latestConsentChangeDate: Date;
+
+  constructor(id: string, name: string) {
```
[the remark this answers is not in the document]
Comment (Nas84) [reply packages/yourloops/lib/auth/user.ts:50]: to keep it simple first. it this need come we will do it.

File: packages/yourloops/lib/auth/user.ts (hunk 7 above)
Comment (Nas84) [reply packages/yourloops/lib/auth/user.ts:91]: yes i know, but for now i did not want to couple the two functions. wanted to wait a little

File: packages/yourloops/lib/auth/api.ts [hunk 16]
```diff
@@ -30,7 +30,8 @@ import bows from "bows";
 import _ from "lodash";
 
 import { APIErrorResponse } from "models/error";
-import { User, Profile, Preferences, Settings, UserRoles } from "../../models/shoreline";
+import User from "./user";
```
[the remark this answers is not in the document]
Comment (Nas84) [reply packages/yourloops/lib/auth/api.ts:33]: i dont understand

File: packages/yourloops/lib/auth/hook.tsx [hunk 17]
```diff
@@ -163,10 +163,10 @@ function AuthContextImpl(api: AuthAPI): AuthContext {
 
     sessionStorage.setItem(STORAGE_KEY_SESSION_TOKEN, auth.sessionToken);
     sessionStorage.setItem(STORAGE_KEY_TRACE_TOKEN, auth.traceToken);
-    sessionStorage.setItem(STORAGE_KEY_USER, JSON.stringify(loggedUser));
+    sessionStorage.setItem(STORAGE_KEY_USER, JSON.stringify(auth.user));
```
[the remark this answers is not in the document]
Comment (Nas84) [reply packages/yourloops/lib/auth/hook.tsx:166]: could you give more explanation ?
why do we need to have a new function when the stringify works?

**Review by yvon-dblg — APPROVED** [review]

**Pull request merged**